staticjinja/staticjinja.py

```diff
@@ -93,39 +117,40 @@ def get_template(self, template_name):
         """
         return self._env.get_template(template_name)
 
-    def _get_context_generator(self, template_name):
-        """Get a context generator for a template.
-
-        Raises a :exc:`ValueError` if no matching context generator can be
-        found.
-
-        :param template_name: the name of the template
-        """
-        for regex, context_generator in self.contexts:
-            if re.match(regex, template_name):
-                return context_generator
-        raise ValueError("no matching context generator")
-
     def get_context(self, template):
         """Get the context for a template.
 
         By default, this function will return an empty context.
 
-        If a matching context_generator can be found, it will be passed the
-        template and executed. The context generator should return a dictionary
-        representing the context.
+        If a matching context_generator can be found and is a function
+        it will be executed after being passed the template if it expects an
+        argument. If it is not a function then it will be assumed to act like
+        a dictionary object.
+
+        If several context_generator can be found, their resulting dictionary
+        will be merged if mergecontexts is True or only the first one will be
+        used otherwise.
 
         :param template: the template to get the context for
         """
-        try:
-            context_generator = self._get_context_generator(template.name)
-        except ValueError:
-            return {}
-        else:
-            try:
-                return context_generator(template)
-            except TypeError:
-                return context_generator()
+
+        context = {}
+        for regex, context_generator in self.contexts:
+            if re.match(regex, template.name):
+                if inspect.isfunction(context_generator):
+                    if _has_argument(context_generator):
+                        context.update(context_generator(template))
+                    else:
+                        context.update(context_generator())
```

Comment:
You have an interesting point about accidentally catching `TypeErrors` in the code we're calling out to, and I'm not sure I can really argue against that being bad. However, I'd argue we should still use the `try-except` style for two reasons. First, it's traditionally Pythonic. Python's philosophy is "it's Easier to Ask Forgiveness than Permission" (EAFP) rather than "Look Before You Leap" (LBYL), which is why I use this style throughout the code. Second, it's much simpler, and doesn't rely on inspection or type checking, with goes along with the philosophy of duck typing. In general, I think `inspect` should only be used as a last resort. Unless you have a strong objection, I suggest we change this back.